pkg/agent/route/route.go

```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
 	ipRule := netlink.NewRule()
 	ipRule.IifName = c.nodeConfig.GatewayConfig.Name
 	ipRule.Mark = iptables.RtTblSelectorValue
+	ipRule.Mask = iptables.RtTblSelectorValue
 	ipRule.Table = AntreaServiceTableIdx
 	ipRule.Priority = AntreaIPRulePriority
 	if err = netlink.RuleDel(ipRule); err != nil {
 		return fmt.Errorf("ip rule delete: %w", err)
 	}
 	return nil
 }
+
+// resolveDefaultRouteNHMAC resolves the MAC of default route next
+// hop on service route table.
+func (c *client) resolveDefaultRouteNHMAC() (net.HardwareAddr, error) {
+	// This MAC is relevant to L2 mode only.
+	// TODO, for now uses AKS default MAC.
+	return net.ParseMAC("12:34:56:78:9a:bc")
+}
+
+// SetupPolicyOnlyMode configures routing needed by traffic policy-only mode.
+func (c *client) SetupPolicyOnlyMode() error {
+	gwLink := util.GetNetLink(c.nodeConfig.GatewayConfig.Name)
+	_, gwIP, _ := net.ParseCIDR(fmt.Sprintf("%s/32", c.nodeConfig.NodeIPAddr.IP.String()))
+	if err := netlink.AddrReplace(gwLink, &netlink.Addr{IPNet: gwIP}); err != nil {
+		klog.Fatalf("Failed to add address %s to gw %s: %v", gwIP, gwLink.Attrs().Name, err)
+	}
+
+	// add default route to service table.
+	_, defaultRt, _ := net.ParseCIDR("0/0")
+	nhIP := net.ParseIP("169.254.253.1")
+	route := &netlink.Route{
+		LinkIndex: gwLink.Attrs().Index,
+		Table:     ServiceRtTable.Idx,
+		Flags:     int(netlink.FLAG_ONLINK),
+		Dst:       defaultRt,
+		Gw:        nhIP,
+	}
+	if err := netlink.RouteReplace(route); err != nil {
+		klog.Fatalf("Failed to add default route to service table: %v", err)
+	}
+	// add static neighbor to next hop so that no ARPING is ever required on gw0
+	nhMAC, _ := c.resolveDefaultRouteNHMAC()
+	neigh := &netlink.Neigh{
+		LinkIndex:    gwLink.Attrs().Index,
+		Family:       netlink.FAMILY_V4,
+		State:        netlink.NUD_PERMANENT,
+		IP:           nhIP,
+		HardwareAddr: nhMAC,
+	}
+	if err := netlink.NeighSet(neigh); err != nil {
+		klog.Fatalf("Failed to add neigh %v to gw %s: %v", neigh, gwLink.Attrs().Name, err)
+	}
+	return nil
+}
+
+// MigrateRoutesToGw moves routes (including assigned IP addresses if any) from link linkName to
+// host gateway.
+func (c *client) MigrateRoutesToGw(linkName string) error {
+	gwLink := util.GetNetLink(c.nodeConfig.GatewayConfig.Name)
+	link, err := netlink.LinkByName(linkName)
+	if err != nil {
+		return fmt.Errorf("failed to get link %s: %w", linkName, err)
+	}
+
+	// Swap route first then address, otherwise route gets removed when address is removed.
+	routes, err := netlink.RouteList(link, netlink.FAMILY_V4)
+	if err != nil {
+		return fmt.Errorf("failed to get routes for link %s: %w", linkName, err)
+	}
+	for _, route := range routes {
+		route.LinkIndex = gwLink.Attrs().Index
+		if err = netlink.RouteReplace(&route); err != nil {
+			return fmt.Errorf("failed to add route %v to link %s: %w", &route, gwLink.Attrs().Name, err)
+		}
+	}
+
+	// swap address if any
```

Comment:
Is there any cloud having addresses on its host side peer devices?

Reply:
some switched CNI solution assigns IP to PtP device in host network, just take care of that case. In term of cloud provider, EKS does not assign host network IP to PtP device.

Reply:
Nit: Swap address if any.